src/riemannian_diff.jl

```diff
@@ -335,3 +335,43 @@ function riemannian_gradient!(
     change_representer!(M, X, embedding_metric, p, X)
     return X
 end
+
+@doc raw"""
+    riemannian_Hessian(M, p, eG, eH, X)
+    riemannian_Hessian!(M, Y, p, eG, eH, X)
+
+Convert the Euclidean Hessian `eH=```\operatorname{Hess} \tilde f(p) [X]``
+of a function ``f \colon \mathcal M \to \mathbb R``, which is the restriction of ``\tilde f`` to ``\mathcal M``,
+given additionally the (Euclidean) gradient ``\operatorname{grad} \tilde f(p)``.
+
+By default it uses the following computation:
+
+Let ``P_Y(X) = D_p(\operatorname{proj}_{T_p\mathcal M}(X)[Y]``, ``Y \in T_o\mathcal M``
+denote the differential of
+the projection onto ``T_p\mathcal M`` with respect to ``p`` (at ``X`` in direction ``Y``) of ``Y`` onto the
+tangent space at ``p``, cf. [`differential_project_basepoint`](@ref)`(M, p, X, Y)`.
+We further denote by ``\operatorname{proj}_{N_p\mathcal M}(X) = X - \operatorname{proj}_{T_p\mathcal M}(X)``
+the projection onto the normal space at ``p``
+
+Then, following Boumal, 2023, Section 5.11 we can compute
+
+```math
+\operatorname{Hess} f(p)[X]
+= \operatorname{proj}_{T_p\mathcal M}\bigl(\operatorname{Hess} \tilde f(p)[X])
++ \mathcal P_X\Bigl( \operatorname{proj}_{N_p\mathcal M}\bigl( \operatorname{grad} \tilde f (p) \bigr) \Bigr)
+```
+
+This method can also be implemented directly, if a more efficient/stable version is known.
+
+The function is inspired by `ehess2rhess` in the [Matlab package Manopt](https://manopt.org).
+"""
+function riemannian_Hessian(M::AbstractManifold, p, eG, eH, X)
+    Y = zero_vector(M, p)
+    riemannian_gradient!(M, Y, p, eG, eH, X)
+    return Y
+end
+function riemannian_Hessian!(M::AbstractManifold, Y, p, eG, eH, X)
+    project!(M, Y, p, eH) #first term
+    Y .= Y + differential_project_basepoint(M, p, eG - project(M, p, eG), X)
```

Comment:
We don't actually need the full differential here, as Boumal's book explains a little further. The restriction that is needed is called "Weingarten map".

Reply:
BTW, it's been known for quite some time, see https://sites.uclouvain.be/absil/2013.01 .

Reply:
Well, this is how far I understood this starting today. And yes I know the reference, but until now I have not yet had the chance to understand that much more. I am also not sure whether this works for any embedding (or whether one needs something like the change_representer for other metrics in the embedding).

Reply:
First of all I am happy to have one thing at least started, that was open for over 4 years. But sure if you see an even nicer generic way, feel free to improve this. I think for several manifolds one would use a closed form for `riemannian_Hessian` anyways. Though I am not sure where to find references for the formulae of `ehess2rhess`from some manifolds in manopt(Matlab); we will see...

Reply:
Oh but sure the II (read. two) operator would be an alternative and using the metric in the embedding (Eq. 5.42). But for neither (`II(X;Y)` nor the current `differential_project_basepoint`) I have a clue how these would look like on any manifold, so then maybe it is irrelevant anyways which formula / function name we document / define?

Reply:
For now I'd suggest just implementing Weingarten maps separately for each embedding and not care about abstractions like `change_representer`. Maybe it will be worked out later, at which point we could refactor our code. So, the only change here I see is to call `differential_project_basepoint` a `weingarten_map` and adjust docs a bit (explain that the argument after `p` is from the normal space at `p`, not any vector in the embedding).

Reply:
But you would need the second fundamental form ``II(M, p, X, Y)`` (5.42) for the shorter form I think, then one can omit a projection onto the normal space even.

Reply:
...and I still feel that differential of the projection is something a larger audience would understand (but I might be biased ;))

Reply:
IMO projecting to normal space is easier then computing `II` :slightly_smiling_face: . Computing the full differential of projection just takes more work than is necessary for Riemannian Hessians. And Weingarten maps seem to be somewhat established in the literature.

Reply:
Actually you should use (5.41) from my copy:
![image](https://github.com/JuliaManifolds/ManifoldDiff.jl/assets/2551062/50561f74-e1bf-4527-8a3e-ed3b0cc9704b)
. I think `weingarten_map(M, p, X, Y)` you described sounds right.

Reply:
Just as I wrote, maybe even define `Weingarten` (maybe even with capital `W` since it is a last name – actually also the wine garden in German so a proper noun as well) in base then.

But sure we can switch to that.

Reply:
OK, that sounds fine.